### ISSUE TITLE
Protocol matching logic doesn't report incompatible use of recursive (self-referential) protocol

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4332]
See https://github.com/python/mypy/issues/14291 for details.

```python
from typing import Protocol, List

class CallTree(Protocol):
    name: str
    subcalls: List["CallTree"]

class MyCallTree:
    name = "myMethod"
    subcalls: List["MyCallTree"]
    
class OtherCallTree:
    name: str
    subcalls: List["CallTree"]

def bad(ct1: CallTree, ct2: CallTree):
    ct1.subcalls.append(ct2)

mct: MyCallTree = MyCallTree()
oct: OtherCallTree = OtherCallTree()
bad(mct, oct)  # now mct.subcalls contains an OtherCallTree, despite being of type list[MyCallTree]
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.
